src/code/PreRender.c

```diff
@@ -1,42 +1,632 @@
-#include <ultra64.h>
 #include <global.h>
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0E70.s")
+void PreRender_SetValuesSave(PreRenderContext* this, u32 width, u32 height, void* fbuf, void* zbuf, void* cvg) {
+    this->widthSave = width;
+    this->heightSave = height;
+    this->fbufSave = fbuf;
+    this->cvgSave = cvg;
+    this->zbufSave = zbuf;
+    this->ulxSave = 0;
+    this->ulySave = 0;
+    this->lrxSave = width - 1;
+    this->lrySave = height - 1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0EA8.s")
+void PreRender_Init(PreRenderContext* this) {
+    bzero(this, sizeof(PreRenderContext));
+    ListAlloc_Init(&this->alloc);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0ED8.s")
+void PreRender_SetValues(PreRenderContext* this, u32 width, u32 height, void* fbuf, void* zbuf) {
+    this->width = width;
+    this->height = height;
+    this->fbuf = fbuf;
+    this->zbuf = zbuf;
+    this->ulx = 0;
+    this->uly = 0;
+    this->lrx = width - 1;
+    this->lry = height - 1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0F08.s")
+void PreRender_Destroy(PreRenderContext* this) {
+    ListAlloc_FreeAll(&this->alloc);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0F28.s")
+void func_800C0F28(PreRenderContext* this, Gfx** dList, void* buf, void* bufSave) {
+    Gfx* dListHead;
+    s32 x;
+    s32 x2;
+    s32 add;
+    s32 uls;
+    s32 ult;
+    s32 lrs;
+    s32 lrt;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C1258.s")
+    LogUtils_CheckNullPointer("this", this, "../PreRender.c", 215);
+    LogUtils_CheckNullPointer("glistpp", dList, "../PreRender.c", 216);
+    dListHead = *dList;
+    LogUtils_CheckNullPointer("glistp", dListHead, "../PreRender.c", 218);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C170C.s")
+    gDPPipeSync(dListHead++);
+    gDPSetOtherMode(dListHead++,
+                    G_AD_PATTERN | G_CD_MAGICSQ | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
+                        G_TD_CLAMP | G_TP_NONE | G_CYC_COPY | G_PM_NPRIMITIVE,
+                    G_AC_NONE | G_ZS_PIXEL | G_RM_NOOP | G_RM_NOOP2);
+    gDPSetColorImage(dListHead++, G_IM_FMT_RGBA, G_IM_SIZ_16b, this->width, bufSave);
+    gDPSetScissor(dListHead++, G_SC_NON_INTERLACE, 0, 0, this->width, this->height);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C1AE8.s")
+    x = this->height;
+    add = 0x1000 / (this->width * 2);
+    uls = 0;
+    x2 = 0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C1B24.s")
+    while (x > 0) {
+        lrs = this->width - 1;
+        if (x < add)
+            add = x;
```

Comment:
clang-tidy should be run on this file to fix style issues like the lack of braces here (`format.sh` runs it along with other things)